src/test/resources/regressions/examples/binary_search_tree.gobra

```diff
@@ -0,0 +1,394 @@
+// Any copyright is dedicated to the Public Domain.
+// http://creativecommons.org/publicdomain/zero/1.0/
+
+// some high-level ideas are inspired by Nagini's iap_bst.py example
+
+package main
+
+type Tree struct {
+    root *node
+}
+
+type node struct {
+  value int
+  left *node
+  right *node
+}
+
+pred (t *Tree) tree() {
+    acc(&t.root) &&
+    (t.root != nil ==> t.root.tree() && t.root.sorted(none[int], none[int]))
+}
+
+pred (n *node) tree() {
+    acc(&n.value) && acc(&n.left) && acc(&n.right) &&
+    (n.left != nil ==> n.left.tree()) &&
+    (n.right != nil ==> n.right.tree())
+}
+
+ghost
+requires n.tree()
+pure func (n *node) sorted(lowerBound, upperBound option[int]) bool {
+    return unfolding n.tree() in (lowerBound != none[int] ==> get(lowerBound) < n.value) &&
+        (upperBound != none[int] ==> n.value < get(upperBound)) &&
+        (n.left != nil ==> n.left.sorted(lowerBound, some(n.value))) &&
+        (n.right != nil ==> n.right.sorted(some(n.value), upperBound))
+}
+
+ghost
+requires n.tree()
+requires n.sorted(oldLowerBound, oldUpperBound)
+requires (oldLowerBound == none[int]) ==> (newLowerBound == none[int])
+requires (oldUpperBound == none[int]) ==> (newUpperBound == none[int])
+requires newLowerBound != none[int] ==> oldLowerBound != none[int] && get(oldLowerBound) >= get(newLowerBound)
+requires newUpperBound != none[int] ==> oldUpperBound != none[int] && get(oldUpperBound) <= get(newUpperBound)
+ensures n.tree()
+ensures n.sorted(oldLowerBound, oldUpperBound) && n.sorted(newLowerBound, newUpperBound)
+ensures n.sortedValues(oldLowerBound, oldUpperBound) == n.sortedValues(newLowerBound, newUpperBound)
+func (n *node) convert(oldLowerBound, oldUpperBound, newLowerBound, newUpperBound option[int]) {
+    unfold n.tree()
+    assert newLowerBound != none[int] ==> get(newLowerBound) < n.value
+    assert newUpperBound != none[int] ==> get(newUpperBound) > n.value
+    if (n.left != nil) {
+        assert n.left.sorted(oldLowerBound, some(n.value))
+        n.left.convert(oldLowerBound, some(n.value), newLowerBound, some(n.value))
+        assert n.left.sorted(newLowerBound, some(n.value))
+    }
+    if (n.right != nil) {
+        assert n.right.sorted(some(n.value), oldUpperBound)
+        n.right.convert(some(n.value), oldUpperBound, some(n.value), newUpperBound)
+        assert n.right.sorted(some(n.value), newUpperBound)
+    }
+    fold n.tree()
+}
+
+ghost
+requires t.tree()
+ensures forall i int :: (0 <= i && i + 1 < len(res) ==> res[i] < res[i + 1]) // ordered
```

Comment:
I am not sure if this post-condition is strong enough to do any "real" proofs with it because it does not ensure that the values that occur in the result list are actually the values that occur in the tree

Reply:
I just tried to naively repeat the pure function's body in the postcondition but this leads to Z3 not terminating. In addition, I've tried to write a postcondition that related the ordered sequence of values to a set of values in the tree but the same problem with Z3 not terminating occurred. Thus, I propose to leave the example as is for the time being